File: src/main/java/dev/daae/time/models/Session.java
package dev.daae.time.models;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import java.time.OffsetDateTime;
import java.util.Optional;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.NonNull;
import lombok.Setter;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder(toBuilder = true)
public class Session {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long id;

  @NonNull private OffsetDateTime start;

  @Column(name = "_end")
  private OffsetDateTime end;

  public Optional<OffsetDateTime> getEnd() {
    return Optional.ofNullable(end);
  }

  public static Session newWithStartTime(OffsetDateTime start) {
    return Session.builder().start(start).build();
  }
}

File: src/main/kotlin/dev/daae/time/models/Session.kt
package dev.daae.time.models

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import java.time.OffsetDateTime

@Entity
class Session(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,
    var start: OffsetDateTime? = null,
    @Column(name = "_end")
    var end: OffsetDateTime? = null,
) {
    companion object {
        @JvmStatic
        fun newWithStartTime(start: OffsetDateTime): Session? {
            return Session(null, start, null)
        }
    }
}
